### ISSUE TITLE
Formatting behaves strangely in Zed

### ISSUE THREAD

**Issue by kwando** [elixir-tools/next-ls#244]
Zed added support for `next-ls` and just took it for a spin. It installed fine, but there is something wrong with the formatting.

If I add a number of new lines inside a function and then saves it formats away the newlines _AND_ it also removes the same number of lines from the top of the file.

Not sure if this is something with Zed or next-ls, elixir ls behaves as expected In Zed though :)

**Comment by mhanberg** [comment]
I had two other reports of this. I'm currently working to see if it's a zed or next problem

**Comment by kwando** [comment]
Not sure where to start digging, let me know if I can help

**Comment by mhanberg** [comment]
I believe there should be a way to open a panel in Zed that shows the RPC messages between the editor and the LSP

You can look at those to see if there is anything peculiar going on.

**Comment by kwando** [comment]
I found the LSP RPC logs :)

1. Clear RPC log
2. Add  newline in the `all/0` function
3. Press save
![Skärmavbild 2023-09-28 kl  15 46 08](https://github.com/elixir-tools/next-ls/assets/891566/c588ee8f-42a9-475e-a830-dff14808e088)


https://github.com/elixir-tools/next-ls/assets/891566/8bd1ea21-d172-43aa-a535-716a130527e1



This is the logs I got from this interaction, first time I see LSP-logs so I do not know what to make from them :)
```JSON
// Send:
{"jsonrpc":"2.0","method":"textDocument/didChange","params":{"textDocument":{"uri":"file:///Users/kwando/projects/sail_race/lib/sail_race/races.ex","version":20},"contentChanges":[{"range":{"start":{"line":4,"character":14},"end":{"line":4,"character":14}},"text":"\n    "}]}}
{"jsonrpc":"2.0","method":"textDocument/didChange","params":{"textDocument":{"uri":"file:///Users/kwando/projects/sail_race/lib/sail_race/races.ex","version":21},"contentChanges":[]}}
{"jsonrpc":"2.0","method":"textDocument/didChange","params":{"textDocument":{"uri":"file:///Users/kwando/projects/sail_race/lib/sail_race/races.ex","version":22},"contentChanges":[{"range":{"start":{"line":5,"character":0},"end":{"line":5,"character":4}},"text":""}]}}
{"jsonrpc":"2.0","id":21,"method":"textDocument/formatting","params":{"textDocument":{"uri":"file:///Users/kwando/projects/sail_race/lib/sail_race/races.ex"},"options":{"tabSize":2,"insertSpaces":true,"insertFinalNewline":true}}}
// Receive:
{"id":21,"jsonrpc":"2.0","result":[{"newText":"","range":{"end":{"character":0,"line":1},"start":{"character":0,"line":0}}}]}
// Send:
{"jsonrpc":"2.0","method":"textDocument/didChange","params":{"textDocument":{"uri":"file:///Users/kwando/projects/sail_race/lib/sail_race/races.ex","version":23},"contentChanges":[{"range":{"start":{"line":0,"character":0},"end":{"line":1,"character":0}},"text":""}]}}
{"jsonrpc":"2.0","method":"textDocument/didSave","params":{"textDocument":{"uri":"file:///Users/kwando/projects/sail_race/lib/sail_race/races.ex"},"text":"  alias SailRace.{DB, Repo}\n  require Ecto.Query\n\n  def all() do\n\n    Ecto.Query.from(r in DB.Race, order_by: r.start_time)\n    |> Repo.all()\n  end\n\n  def all_completed() do\n    Ecto.Query.from(r in DB.Race, order_by: r.start_time, where: r.completed)\n    |> Repo.all()\n  end\n\n  def find(race_id) do\n    Repo.find(DB.Race, race_id)\n  end\n\n  def create_race(params) do\n    %DB.Race{}\n    |> DB.Race.changeset(Map.new(params))\n    |> Repo.insert()\n  end\n\n  def update(race, params) do\n    race\n    |> DB.Race.changeset(Map.new(params))\n    |> Repo.update()\n  end\n\n  def next_race() do\n    Ecto.Query.from(r in DB.Race,\n      where: not r.completed and r.start_time >= ^DateTime.utc_now(),\n      order_by: r.start_time,\n      limit: 1\n    )\n    |> Repo.one()\n  end\n\n  def create_from_start_times(start_times) do\n    Repo.transaction(fn ->\n      for start_time <- start_times do\n        DB.Race.changeset(%DB.Race{}, %{start_time: start_time})\n        |> Repo.insert!()\n      end\n    end)\n  end\nend\n"}}
// Receive:
{"jsonrpc":"2.0","method":"$/progress","params":{"token":"XV1hD4W3","value":{"kind":"begin","title":"Compiling ..."}}}
{"jsonrpc":"2.0","method":"window/logMessage","params":{"message":"[NextLS] Compiling 1 file (.ex)","type":4}}
{"jsonrpc":"2.0","method":"window/logMessage","params":{"message":"[NextLS] \n== Compilation error in file lib/sail_race/races.ex ==\n** (SyntaxError) lib/sail_race/races.ex:48:1: unexpected reserved word: end\n    |\n 48 | end\n    | ^\n    (elixir 1.15.6) lib/kernel/parallel_compiler.ex:377: anonymous fn/5 in Kernel.ParallelCompiler.spawn_workers/8","type":4}}
{"jsonrpc":"2.0","method":"window/logMessage","params":{"message":"[NextLS] Compiled !","type":4}}
{"jsonrpc":"2.0","method":"textDocument/publishDiagnostics","params":{"diagnostics":[{"data":null,"message":"** (SyntaxError) lib/sail_race/races.ex:48:1: unexpected reserved word: end\n    |\n 48 | end\n    | ^\n    (elixir 1.15.6) lib/kernel/parallel_compiler.ex:377: anonymous fn/5 in Kernel.ParallelCompiler.spawn_workers/8\n","range":{"end":{"character":999,"line":47},"start":{"character":0,"line":47}},"severity":1,"source":"Elixir"}],"uri":"file:///Users/kwando/projects/sail_race/lib/sail_race/races.ex"}}
{"jsonrpc":"2.0","method":"$/progress","params":{"token":"XV1hD4W3","value":{"kind":"end","message":"Compiled !"}}}
{"jsonrpc":"2.0","method":"textDocument/publishDiagnostics","params":{"diagnostics":[{"data":null,"message":"** (SyntaxError) lib/sail_race/races.ex:48:1: unexpected reserved word: end\n    |\n 48 | end\n    | ^\n    (elixir 1.15.6) lib/kernel/parallel_compiler.ex:377: anonymous fn/5 in Kernel.ParallelCompiler.spawn_workers/8\n","range":{"end":{"character":999,"line":47},"start":{"character":0,"line":47}},"severity":1,"source":"Elixir"}],"uri":"file:///Users/kwando/projects/sail_race/lib/sail_race/races.ex"}}
```

Server logs
```
[NextLS] NextLS v0.12.6 has initialized!
[NextLS] Booting runtimes...
[NextLS] Beginning DB migration...
[NextLS] Database is on the latest version: 5
[NextLS] Finished DB migration...
[NextLS] Connected to node nextls-runtime-1695908620235890041@eve
[NextLS] Runtime for folder  is ready...
[NextLS] Compiled !
```

**Comment by kwando** [comment]
The first receive looks a pit peculiar though, must be this one that causes the line to disappear 
```
{"id":21,"jsonrpc":"2.0","result":[{"newText":"","range":{"end":{"character":0,"line":1},"start":{"character":0,"line":0}}}]}
```

**Comment by mhanberg** [comment]
can you do the same process, but open a new file after clearing the logs? it should be sending a `didOpen` message and i want to see what's in there.

but interestingly, they are doing format -> save, rather than save -> format. i think that might be a clue

**Comment by kwando** [comment]
This is the same thing but I opened the file first :)

```json
// Send:
{"jsonrpc":"2.0","method":"textDocument/didOpen","params":{"textDocument":{"uri":"file:///Users/kwando/projects/sail_race/lib/sail_race/boats.ex","languageId":"","version":0,"text":"defmodule SailRace.Boats do\n  alias SailRace.{DB, Repo}\n  require Ecto.Query\n\n  def all() do\n    Ecto.Query.from(b in DB.Boat)\n    |> Repo.all()\n  end\n\n  def without_result_in_race(race_id) do\n    Ecto.Query.from(b in DB.Boat,\n      left_join: rr in DB.RaceResult,\n      on: rr.boat_id == b.id and rr.race_id == ^race_id,\n      where: is_nil(rr.id)\n    )\n    |> Repo.all()\n  end\n\n  def find(boat_id), do: Repo.find(DB.Boat, boat_id)\n\n  def create(params) do\n    %DB.Boat{}\n    |> DB.Boat.changeset(params)\n    |> Repo.insert()\n  end\n\n  def update(boat_id, params) do\n    {:ok, boat} = find(boat_id)\n\n    boat\n    |> DB.Boat.changeset(params)\n    |> Repo.update()\n  end\n\n  def delete(boat) do\n    Repo.delete(boat)\n    :ok\n  end\n\n  @mapping %{\n    \"signup_id\" => \"id\",\n    \"skipper\" => \"skipper\",\n    \"model\" => \"boat_model\",\n    \"sail_number\" => \"sail_number\",\n    \"name\" => \"boat_name\",\n    \"club_name\" => \"club\"\n  }\n  def import_unregistered_boats() do\n    {:ok, signups} = SignupAPI.get_signups()\n\n    signup_ids = MapSet.new(for signup <- signups, do: signup[\"id\"])\n\n    existing_signups =\n      Ecto.Query.from(b in DB.Boat, select: b.signup_id, where: not is_nil(b.signup_id))\n      |> Repo.all()\n      |> MapSet.new()\n\n    new_signups = MapSet.difference(signup_ids, existing_signups)\n\n    for signup <- signups do\n      if MapSet.member?(new_signups, signup[\"id\"]) do\n        params =\n          for {key, from} <- @mapping, into: %{} do\n            {key, Map.fetch!(signup, from)}\n          end\n\n        IO.inspect(signup, label: \"IMPORT\")\n\n        create(params)\n        |> IO.inspect()\n      end\n    end\n\n    :ok\n  end\n\n  def delete_imported! do\n    Ecto.Query.from(b in DB.Boat, where: not is_nil(b.signup_id))\n    |> Repo.delete_all()\n  end\nend\n"}}}
{"jsonrpc":"2.0","method":"textDocument/didChange","params":{"textDocument":{"uri":"file:///Users/kwando/projects/sail_race/lib/sail_race/boats.ex","version":1},"contentChanges":[{"range":{"start":{"line":4,"character":14},"end":{"line":4,"character":14}},"text":"\n    "}]}}
{"jsonrpc":"2.0","method":"textDocument/didChange","params":{"textDocument":{"uri":"file:///Users/kwando/projects/sail_race/lib/sail_race/boats.ex","version":2},"contentChanges":[]}}
{"jsonrpc":"2.0","method":"textDocument/didChange","params":{"textDocument":{"uri":"file:///Users/kwando/projects/sail_race/lib/sail_race/boats.ex","version":3},"contentChanges":[{"range":{"start":{"line":5,"character":0},"end":{"line":5,"character":4}},"text":""}]}}
{"jsonrpc":"2.0","id":36,"method":"textDocument/formatting","params":{"textDocument":{"uri":"file:///Users/kwando/projects/sail_race/lib/sail_race/boats.ex"},"options":{"tabSize":2,"insertSpaces":true,"insertFinalNewline":true}}}
// Receive:
{"id":36,"jsonrpc":"2.0","result":[{"newText":"","range":{"end":{"character":0,"line":1},"start":{"character":0,"line":0}}}]}
// Send:
{"jsonrpc":"2.0","method":"textDocument/didChange","params":{"textDocument":{"uri":"file:///Users/kwando/projects/sail_race/lib/sail_race/boats.ex","version":4},"contentChanges":[{"range":{"start":{"line":0,"character":0},"end":{"line":1,"character":0}},"text":""}]}}
{"jsonrpc":"2.0","method":"textDocument/didSave","params":{"textDocument":{"uri":"file:///Users/kwando/projects/sail_race/lib/sail_race/boats.ex"},"text":"  alias SailRace.{DB, Repo}\n  require Ecto.Query\n\n  def all() do\n\n    Ecto.Query.from(b in DB.Boat)\n    |> Repo.all()\n  end\n\n  def without_result_in_race(race_id) do\n    Ecto.Query.from(b in DB.Boat,\n      left_join: rr in DB.RaceResult,\n      on: rr.boat_id == b.id and rr.race_id == ^race_id,\n      where: is_nil(rr.id)\n    )\n    |> Repo.all()\n  end\n\n  def find(boat_id), do: Repo.find(DB.Boat, boat_id)\n\n  def create(params) do\n    %DB.Boat{}\n    |> DB.Boat.changeset(params)\n    |> Repo.insert()\n  end\n\n  def update(boat_id, params) do\n    {:ok, boat} = find(boat_id)\n\n    boat\n    |> DB.Boat.changeset(params)\n    |> Repo.update()\n  end\n\n  def delete(boat) do\n    Repo.delete(boat)\n    :ok\n  end\n\n  @mapping %{\n    \"signup_id\" => \"id\",\n    \"skipper\" => \"skipper\",\n    \"model\" => \"boat_model\",\n    \"sail_number\" => \"sail_number\",\n    \"name\" => \"boat_name\",\n    \"club_name\" => \"club\"\n  }\n  def import_unregistered_boats() do\n    {:ok, signups} = SignupAPI.get_signups()\n\n    signup_ids = MapSet.new(for signup <- signups, do: signup[\"id\"])\n\n    existing_signups =\n      Ecto.Query.from(b in DB.Boat, select: b.signup_id, where: not is_nil(b.signup_id))\n      |> Repo.all()\n      |> MapSet.new()\n\n    new_signups = MapSet.difference(signup_ids, existing_signups)\n\n    for signup <- signups do\n      if MapSet.member?(new_signups, signup[\"id\"]) do\n        params =\n          for {key, from} <- @mapping, into: %{} do\n            {key, Map.fetch!(signup, from)}\n          end\n\n        IO.inspect(signup, label: \"IMPORT\")\n\n        create(params)\n        |> IO.inspect()\n      end\n    end\n\n    :ok\n  end\n\n  def delete_imported! do\n    Ecto.Query.from(b in DB.Boat, where: not is_nil(b.signup_id))\n    |> Repo.delete_all()\n  end\nend\n"}}
// Receive:
{"jsonrpc":"2.0","method":"$/progress","params":{"token":"VMqCmp7b","value":{"kind":"begin","title":"Compiling ..."}}}
{"jsonrpc":"2.0","method":"window/logMessage","params":{"message":"[NextLS] Compiling 1 file (.ex)","type":4}}
{"jsonrpc":"2.0","method":"window/logMessage","params":{"message":"[NextLS] \n== Compilation error in file lib/sail_race/boats.ex ==\n** (SyntaxError) lib/sail_race/boats.ex:81:1: unexpected reserved word: end\n    |\n 81 | end\n    | ^\n    (elixir 1.15.6) lib/kernel/parallel_compiler.ex:377: anonymous fn/5 in Kernel.ParallelCompiler.spawn_workers/8","type":4}}
{"jsonrpc":"2.0","method":"window/logMessage","params":{"message":"[NextLS] Compiled !","type":4}}
{"jsonrpc":"2.0","method":"$/progress","params":{"token":"VMqCmp7b","value":{"kind":"end","message":"Compiled !"}}}
{"jsonrpc":"2.0","method":"textDocument/publishDiagnostics","params":{"diagnostics":[{"data":null,"message":"** (SyntaxError) lib/sail_race/boats.ex:81:1: unexpected reserved word: end\n    |\n 81 | end\n    | ^\n    (elixir 1.15.6) lib/kernel/parallel_compiler.ex:377: anonymous fn/5 in Kernel.ParallelCompiler.spawn_workers/8\n","range":{"end":{"character":999,"line":80},"start":{"character":0,"line":80}},"severity":1,"source":"Elixir"}],"uri":"file:///Users/kwando/projects/sail_race/lib/sail_race/boats.ex"}}
{"jsonrpc":"2.0","method":"textDocument/publishDiagnostics","params":{"diagnostics":[],"uri":"file:///Users/kwando/projects/sail_race/lib/sail_race/races.ex"}}
{"jsonrpc":"2.0","method":"textDocument/publishDiagnostics","params":{"diagnostics":[{"data":null,"message":"** (SyntaxError) lib/sail_race/boats.ex:81:1: unexpected reserved word: end\n    |\n 81 | end\n    | ^\n    (elixir 1.15.6) lib/kernel/parallel_compiler.ex:377: anonymous fn/5 in Kernel.ParallelCompiler.spawn_workers/8\n","range":{"end":{"character":999,"line":80},"start":{"character":0,"line":80}},"severity":1,"source":"Elixir"}],"uri":"file:///Users/kwando/projects/sail_race/lib/sail_race/boats.ex"}}
{"jsonrpc":"2.0","method":"textDocument/publishDiagnostics","params":{"diagnostics":[],"uri":"file:///Users/kwando/projects/sail_race/lib/sail_race/races.ex"}}
```

**Comment by kwando** [comment]
This is the elixir-ls is doing when I follow the same process
```
// Send:
{"jsonrpc":"2.0","method":"textDocument/didChange","params":{"textDocument":{"uri":"file:///Users/kwando/projects/sail_race/lib/sail_race/races.ex","version":5},"contentChanges":[{"range":{"start":{"line":4,"character":14},"end":{"line":4,"character":14}},"text":"\n  "}]}}
{"jsonrpc":"2.0","method":"textDocument/didChange","params":{"textDocument":{"uri":"file:///Users/kwando/projects/sail_race/lib/sail_race/races.ex","version":6},"contentChanges":[{"range":{"start":{"line":5,"character":0},"end":{"line":5,"character":0}},"text":"  "}]}}
{"jsonrpc":"2.0","method":"textDocument/didChange","params":{"textDocument":{"uri":"file:///Users/kwando/projects/sail_race/lib/sail_race/races.ex","version":7},"contentChanges":[{"range":{"start":{"line":5,"character":0},"end":{"line":5,"character":4}},"text":""}]}}
{"jsonrpc":"2.0","id":3,"method":"textDocument/formatting","params":{"textDocument":{"uri":"file:///Users/kwando/projects/sail_race/lib/sail_race/races.ex"},"options":{"tabSize":2,"insertSpaces":true,"insertFinalNewline":true}}}
// Receive:
{"id":3,"jsonrpc":"2.0","result":[{"range":{"end":{"character":0,"line":6},"start":{"character":0,"line":5}},"newText":""}]}
// Send:
{"jsonrpc":"2.0","method":"textDocument/didChange","params":{"textDocument":{"uri":"file:///Users/kwando/projects/sail_race/lib/sail_race/races.ex","version":8},"contentChanges":[{"range":{"start":{"line":5,"character":0},"end":{"line":6,"character":0}},"text":""}]}}
{"jsonrpc":"2.0","method":"textDocument/didSave","params":{"textDocument":{"uri":"file:///Users/kwando/projects/sail_race/lib/sail_race/races.ex"},"text":"defmodule SailRace.Races do\n  alias SailRace.{DB, Repo}\n  require Ecto.Query\n\n  def all() do\n    Ecto.Query.from(r in DB.Race, order_by: r.start_time)\n    |> Repo.all()\n  end\n\n  def all_completed() do\n    Ecto.Query.from(r in DB.Race, order_by: r.start_time, where: r.completed)\n    |> Repo.all()\n  end\n\n  def find(race_id) do\n    Repo.find(DB.Race, race_id)\n  end\n\n  def create_race(params) do\n    %DB.Race{}\n    |> DB.Race.changeset(Map.new(params))\n    |> Repo.insert()\n  end\n\n  def update(race, params) do\n    race\n    |> DB.Race.changeset(Map.new(params))\n    |> Repo.update()\n  end\n\n  def next_race() do\n    Ecto.Query.from(r in DB.Race,\n      where: not r.completed and r.start_time >= ^DateTime.utc_now(),\n      order_by: r.start_time,\n      limit: 1\n    )\n    |> Repo.one()\n  end\n\n  def create_from_start_times(start_times) do\n    Repo.transaction(fn ->\n      for start_time <- start_times do\n        DB.Race.changeset(%DB.Race{}, %{start_time: start_time})\n        |> Repo.insert!()\n      end\n    end)\n  end\nend\n"}}
// Receive:
{"jsonrpc":"2.0","method":"window/logMessage","params":{"message":"Starting build with MIX_ENV: test MIX_TARGET: host","type":3}}
{"jsonrpc":"2.0","method":"window/logMessage","params":{"message":"Compile took 682 milliseconds","type":3}}
{"jsonrpc":"2.0","method":"textDocument/publishDiagnostics","params":{"diagnostics":[{"message":"The pattern can never match the type.\n\nPattern:\n:prod\n\nType:\n:test\n","range":{"end":{"character":25,"line":8},"start":{"character":2,"line":8}},"severity":2,"source":"ElixirLS Dialyzer"}],"uri":"file:///Users/kwando/projects/sail_race/lib/sail_race/db/supervisor.ex"}}
{"jsonrpc":"2.0","method":"window/logMessage","params":{"message":"[ElixirLS WorkspaceSymbols] Updating index...","type":3}}
{"jsonrpc":"2.0","method":"window/logMessage","params":{"message":"[ElixirLS Dialyzer] Checking for stale beam files","type":3}}
{"jsonrpc":"2.0","method":"window/logMessage","params":{"message":"[ElixirLS Dialyzer] Found 0 changed files in 138 milliseconds","type":3}}
{"jsonrpc":"2.0","method":"window/logMessage","params":{"message":"[ElixirLS Dialyzer] Analyzing 0 modules: []","type":3}}
{"jsonrpc":"2.0","method":"window/logMessage","params":{"message":"[ElixirLS Dialyzer] Analysis finished in 15 milliseconds","type":3}}
{"jsonrpc":"2.0","method":"textDocument/publishDiagnostics","params":{"diagnostics":[{"message":"The pattern can never match the type.\n\nPattern:\n:prod\n\nType:\n:test\n","range":{"end":{"character":25,"line":8},"start":{"character":2,"line":8}},"severity":2,"source":"ElixirLS Dialyzer"}],"uri":"file:///Users/kwando/projects/sail_race/lib/sail_race/db/supervisor.ex"}}
{"jsonrpc":"2.0","method":"window/logMessage","params":{"message":"Dialyzer analysis is up to date","type":3}}
{"jsonrpc":"2.0","method":"window/logMessage","params":{"message":"[ElixirLS WorkspaceSymbols] 1 modules need reindexing","type":3}}
{"jsonrpc":"2.0","method":"window/logMessage","params":{"message":"[ElixirLS WorkspaceSymbols] 0 callbacks added to index","type":3}}
{"jsonrpc":"2.0","method":"window/logMessage","params":{"message":"[ElixirLS WorkspaceSymbols] 0 types added to index","type":3}}
{"jsonrpc":"2.0","method":"window/logMessage","params":{"message":"[ElixirLS WorkspaceSymbols] 1 modules added to index","type":3}}
{"jsonrpc":"2.0","method":"window/logMessage","params":{"message":"[ElixirLS WorkspaceSymbols] 10 functions added to index","type":3}}
{"jsonrpc":"2.0","method":"window/logMessage","params":{"message":"[ElixirLS Dialyzer] Writing manifest...","type":3}}
{"jsonrpc":"2.0","method":"window/logMessage","params":{"message":"[ElixirLS Dialyzer] Done writing manifest in 735 milliseconds.","type":3}}
```

**Comment by mhanberg** [comment]
There's a chance the server is doing something wrong with regard to the spec. 

I'll be able to check it out this weeken if no one else checks it out first. 

Thanks for the reconnaissance!

**Comment by kwando** [comment]
No hurry :)

**Comment by mhanberg** [comment]
I think this is the issue: https://github.com/zed-industries/community/issues/2098

**Comment by kwando** [comment]
Wow, good catch! Thats sounds very plausible.

Maybe `next-ls` should send an error back if this happens instead of blindly accepting it?

**Comment by mhanberg** [comment]
Should be fixed in latest zed release 0.106.2

https://zed.dev/releases/preview

**Comment by kwando** [comment]
Thanks for checking this out!

Have a great weekend 🍻
